fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+	hints->mode = FI_CONTEXT2;
+	hints->ep_attr->type = FI_EP_RDM;
+	hints->tx_attr->op_flags = FI_DELIVERY_COMPLETE;
+
+	ret = fi_getinfo(FI_VERSION(1,6), node, service, flags, hints, &info);
+	if (ret)
+		goto err_getinfo;
+
+	ret = fi_fabric(info->fabric_attr, &ofi_state->fabric, context);
+	if (ret) {
+		hpcs_error("fi_fabric failed\n");
+		goto err_fabric;
+	}
+
+	ret = init_domain(arguments, test_config, ofi_state->fabric, info,
+			ofi_state->domain_state, job);
+	if (ret) {
+		hpcs_error("init_domain failed\n");
+		goto err_domain;
+	} else {
+		hpcs_verbose("init_domain complete, Using %s provider\n",
+			     info->fabric_attr->name);
+	}
+
+	return 0;
+
+err_domain:
+	fi_close(&ofi_state->fabric->fid);
+
+err_fabric:
+	fi_freeinfo(info);
+
+err_getinfo:
+	fi_freeinfo(hints);
+
+	return ret;
+}
+
+static int fini_completion(struct domain_state *domain_state)
+{
+	int ret;
+
+	if (domain_state->tx_cntr != NULL) {
+		ret = fi_close(&domain_state->tx_cntr->fid);
+		if (ret) {
+			hpcs_error("unable to close tx counter\n");
+			return -1;
+		}
+		domain_state->tx_cntr = NULL;
+	}
+
+	if (domain_state->rx_cntr != NULL) {
+		ret = fi_close(&domain_state->rx_cntr->fid);
+		if (ret) {
+			hpcs_error("unable to close rx counter\n");
+			return -1;
+		}
+		domain_state->rx_cntr = NULL;
+	}
+
+	if (domain_state->tx_cq != NULL) {
+		ret = fi_close(&domain_state->tx_cq->fid);
+		if (ret) {
+			hpcs_error("unable to close tx cq\n");
+			return -1;
+		}
+		domain_state->tx_cq = NULL;
+	}
+
+	if (domain_state->rx_cq != NULL) {
+		ret = fi_close(&domain_state->rx_cq->fid);
+		if (ret) {
+			hpcs_error("unable to close rx cq\n");
+			return -1;
+		}
+		domain_state->rx_cq = NULL;
+	}
+
+	return ret;
+}
+
+static int fini_domain(struct domain_state *domain_state)
+{
+	int ret = 0;
+
+	ret = fi_close(&domain_state->endpoint->fid);
+	if (ret) {
+		hpcs_error("unable to close endpoint\n");
+		return ret;
+	}
+	domain_state->endpoint = NULL;
+
+	ret = fi_close(&domain_state->av->fid);
+	if (ret) {
+		hpcs_error("unable to close address vector\n");
+		return ret;
+	}
+
+	ret = fini_completion(domain_state);
+	if (ret)
+		return ret;
+
+	ret = fi_close(&domain_state->domain->fid);
+	if (ret) {
+		hpcs_error("unable to close domain\n");
+		return ret;
+	}
+	domain_state->domain = NULL;
+
+	return 0;
+}
+
+/*
+ * Finalizes OFI.  Safely cleans up resources created by init_ofi.  Takes a
+ * "best effort" approach - goes as far as it can and returns the first error
+ * encountered.
+ */
+int fini_ofi(struct ofi_state *ofi_state)
+{
+	int ret;
+
+	ret = fini_domain(ofi_state->domain_state);
+	if (ret)
+		return ret;
+
+	ofi_state->domain_state = NULL;
+
+	ret = fi_close(&ofi_state->fabric->fid);
+	if (ret) {
+		hpcs_error("unable to close fabric\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * HPCS uses groups of arguments separated by a "-" or "--" (which are treated
+ * as special by getopt, and cause it to stop parsing arguments).
```

Comment:
Why be non-compliant with getopt?

Reply:
We wanted to be able to define tests and patterns independently and not have to deal with a single massive getopt switch somewhere that handles all the arguments, and also not have to worry about argument name collisions.  (Sometimes name collisions are actually rather convenient; for instance being able to define "--help" and "--verbose" as valid arguments to core, or to any pattern, or to any test.)